services/galley/src/Galley/API/Update.hs

```diff
@@ -439,9 +441,14 @@ withValidOtrBroadcastRecipients usr clt rcps val now go = Teams.withBindingTeam
     tMembers <- fmap (view userId) <$> Data.teamMembers tid
     contacts <- getContactList usr
     let users = Set.toList $ Set.union (Set.fromList tMembers) (Set.fromList contacts)
-    clts  <- Data.lookupClients users
+    clts  <- lookupClientsAsync users
     let membs = Data.newMember <$> users
     handleOtrResponse usr clt rcps membs clts val now go
+  where
+    lookupClientsAsync users = do
+        -- do a maximum of 16 parallel lookups of up to 128 users each
+        cltsList  <- forM (chunksOf 2048 users) $ \us -> mapConcurrently Data.lookupClients (chunksOf 128 us)
+        return $ Clients.fromList $ concat $ Clients.toList <$> concat cltsList
```

Comment:
is it necessary going to and from list here? isn't this an instance of Monoid?

Reply:
PR was updated - before, the type was `[[Clients]]` , Clients being a wrapper around `Map UserId [ClientId]`. One can't concat the outer list with the inner map directly, and the return type needs to be of `Clients`. But the updated PR version no longer has a back-and-forth.

Reply:
ok. I thought it was this type:
https://github.com/wireapp/wire-server/blob/develop/services/galley/src/Galley/Types/Clients.hs#L34
which we added a Monoid derivation for recently. I am not sure how mconcat is actually derived though, but I assume it uses mconcat of the inner structures.

It seems it's irrelevant now anyway :)